app/src/behaviors/behavior_macro.c

```diff
@@ -110,21 +137,54 @@ static int behavior_macro_init(const struct device *dev) {
     return 0;
 };
 
+static void replace_params(struct behavior_macro_trigger_state *state,
+                           struct zmk_behavior_binding *binding,
+                           const struct zmk_behavior_binding *macro_binding) {
+    switch (state->param1_source) {
+    case PARAM_SOURCE_MACRO_1ST:
+        binding->param1 = macro_binding->param1;
+        break;
+    case PARAM_SOURCE_MACRO_2ND:
+        binding->param1 = macro_binding->param2;
+        break;
+    default:
+        break;
+    }
+
+    switch (state->param2_source) {
+    case PARAM_SOURCE_MACRO_1ST:
+        binding->param2 = macro_binding->param1;
+        break;
+    case PARAM_SOURCE_MACRO_2ND:
+        binding->param2 = macro_binding->param2;
+        break;
+    default:
+        break;
+    }
```

Comment:
This could be deduplicated with something like

```c
    update_param(&binding->param1, state->param1_source, &macro_binding);
    update_param(&binding->param2, state->param2_source, &macro_binding);
```

Or more functional programming style:

```c
    binding->param1 = update_param(state->param1_source, binding->param1, &macro_binding);
    binding->param2 = update_param(state->param2_source, binding->param2, &macro_binding);
```

Reply:
Good catch. Used the more functional style.